File: algorithm/leetcodeCppTmp/remove-invalid-parentheses.cpp
301. Remove Invalid Parentheses
Difficulty: Hard

Remove the minimum number of invalid parentheses in order to make the input string valid.
Return all possible results.

Note: The input string may contain letters other than the parentheses ( and ).

Examples:
"()())()" -> ["()()()", "(())()"]
"(a)())()" -> ["(a)()()", "(a())()"]
")(" -> [""]

// Time:  O(C(n, c)), try out all possible substrings with the minimum c deletion.
// Space: O(c), the depth is at most c, and it costs n at each depth

// DFS solution with removed array. (4ms)
class Solution {
public:
    vector<string> removeInvalidParentheses(string s)  {
        class Solution {
public:
    vector<string> removeInvalidParentheses(string s) {
        unordered_set<string> visited;
        bool found = false;
        vector<string> result;
        queue<string> q;
        q.push(s);
        visited.insert(s);

        while (!q.empty()) {
            s = q.front(); q.pop();
            if (isValid(s)) {
                result.push_back(s);
                found = true;
            }
            if (found) {
                continue;
            }
            for (int i = 0; i < s.size(); ++i) {
                if (s[i] != '(' && s[i] != ')') {
                    continue;
                }    
                string subS = s.substr(0, i) + s.substr(i + 1);
                if (!visited.count(subS)) {
                    q.push(subS);
                    visited.insert(subS);
                }
            }
        }
        return result;
    }
    bool isValid(string t) {
        int cnt = 0;
        for (int i = 0; i < t.size(); ++i) {
            if (t[i] == '(') {
                ++cnt;
            }
            if (t[i] == ')' && cnt-- == 0) {
                return false;
            }
        }
        return cnt == 0;
    }
};

File: data-structure-and-algorithm/javascript/leetcode/remove-invalid-parentheses.js
/*
Given a string s that contains parentheses and letters, remove the minimum number of invalid parentheses to make the input string valid.

Return all the possible results. You may return the answer in any order.

 

Example 1:

Input: s = "()())()"
Output: ["(())()","()()()"]
Example 2:

Input: s = "(a)())()"
Output: ["(a())()","(a)()()"]
Example 3:

Input: s = ")("
Output: [""]
 

Constraints:

1 <= s.length <= 25
s consists of lowercase English letters and parentheses '(' and ')'.
There will be at most 20 parentheses in s.
*/

function removeInvalidParentheses(s) {
    const result = [];
    dfs(s, 0, 0, ['(', ')'], result);
    return result;   
}

function dfs(s, i0, j0, pair, result) {
    let cnt = 0;
    for (let i = i0; i < s.length; ++i) {
        if (s[i] === pair[0]) {
            ++cnt;
        } else if (s[i] === pair[1]) {
            --cnt;
        }
        if (cnt < 0) {
            for (let j = j0; j <= i; ++j) {
                if (s[j] === pair[1] && (j === j0 || s[j] !== s[j - 1])) {
                    dfs(s.substring(0, j) + s.substring(j + 1), i, j, pair, result);
                }
            }
            return;
        }
    }
    const rs = [...s].reverse().join("");
    if (pair[0] === "(") {
        dfs(rs, 0, 0, [")", "("], result);
    } else {
        result.push(rs);
    }
}